File: roms/jr_pac-man.py
#
#	Jr. Pac-Man
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg = z.read('jrp8d.8d') + z.read('jrp8e.8e') + z.read('jrp8h.8h') + z.read('jrp8j.8j') + z.read('jrp8k.8k')
    bg = z.read('jrp2c.2c')
    obj = z.read('jrp2e.2e')
    rgb_l = z.read('a290-27axv-bxhd.9e')
    rgb_h = z.read('a290-27axv-cxhd.9f')
    color = z.read('a290-27axv-axhd.9p')
    snd = z.read('a290-27axv-dxhd.7p')

rom = prg + bg + obj + rgb_l + rgb_h + color + snd

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/jr_pac-man.js
/*
 *	Jr. Pac-Man
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg = cat(z.read('jrp8d.8d'), z.read('jrp8e.8e'), z.read('jrp8h.8h'), z.read('jrp8j.8j'), z.read('jrp8k.8k'));
const bg = z.read('jrp2c.2c');
const obj = z.read('jrp2e.2e');
const rgb_l = z.read('a290-27axv-bxhd.9e');
const rgb_h = z.read('a290-27axv-cxhd.9f');
const color = z.read('a290-27axv-axhd.9p');
const snd = z.read('a290-27axv-dxhd.7p');

const rom = cat(prg, bg, obj, rgb_l, rgb_h, color, snd);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
